Net/SMTP.php

```diff
@@ -568,7 +568,58 @@ protected function getBestAuthMethod()
 
         return PEAR::raiseError('No supported authentication methods');
     }
-
+    
+    /**
+     * Establish STARTTLS Connection.
+     *
+     * @return mixed Returns a PEAR_Error with an error message on any
+     *               kind of failure, or true on success.
+     * @since 1.9.3
+     */
+    public function starttls()
+    {
+        /* We can only attempt a TLS connection if one has been requested,
+         * we're running PHP 5.1.0 or later, have access to the OpenSSL
+         * extension, are connected to an SMTP server which supports the
+         * STARTTLS extension, and aren't already connected over a secure
+         * (SSL) socket connection. */
+        if (version_compare(PHP_VERSION, '5.1.0', '>=')
+            && extension_loaded('openssl') && isset($this->esmtp['STARTTLS'])
+            && strncasecmp($this->host, 'ssl://', 6) !== 0
+            ) {
+                /* Start the TLS connection attempt. */
+                if (PEAR::isError($result = $this->put('STARTTLS'))) {
+                    return $result;
+                }
+                if (PEAR::isError($result = $this->parseResponse(220))) {
+                    return $result;
+                }
+                if (isset($this->socket_options['ssl']['crypto_method'])) {
+                    $crypto_method = $this->socket_options['ssl']['crypto_method'];
+                } else {
+                    /* STREAM_CRYPTO_METHOD_TLS_ANY_CLIENT constant does not exist
+                     * and STREAM_CRYPTO_METHOD_SSLv23_CLIENT constant is
+                     * inconsistent across PHP versions. */
+                    $crypto_method = STREAM_CRYPTO_METHOD_TLS_CLIENT
+                    | @STREAM_CRYPTO_METHOD_TLSv1_1_CLIENT
+                    | @STREAM_CRYPTO_METHOD_TLSv1_2_CLIENT;
+                }
+                if (PEAR::isError($result = $this->socket->enableCrypto(true, $crypto_method))) {
+                    return $result;
+                } elseif ($result !== true) {
+                    return PEAR::raiseError('STARTTLS failed');
+                }
+                
+                /* Send EHLO again to recieve the AUTH string from the
+                 * SMTP server. */
+                $this->negotiate();
+            } else {
+                return PEAR::raiseError('STARTTLS failed');
```

Comment:
In the previous code, if one of the above conditions failed, we simply wouldn't attempt TLS. In this version, the code now raises an error.

For compatibility with the previous logic, this `else` should be removed so we fall through to the default `return true;` result.

Reply:
I think it would also be fine to `return false;` from this path (instead of raising the error) to indicate that TLS wasn't successful (but also didn't result in an error).